crates/corro-agent/src/agent/tests.rs

```diff
@@ -903,6 +908,106 @@ async fn test_clear_empty_versions() -> eyre::Result<()> {
+    let ta1 = launch_test_agent(|conf| conf.build(), tripwire.clone()).await?;
+    let actor_id = ActorId(Uuid::new_v4());
+    // setup the schema, for both nodes
+    let (status_code, _body) = api_v1_db_schema(
+        Extension(ta1.agent.clone()),
+        axum::Json(vec![corro_tests::TEST_SCHEMA.into()]),
+    )
+    .await;
+    assert_eq!(status_code, StatusCode::OK);
+
+    let ta2 = launch_test_agent(|conf| conf.build(), tripwire.clone()).await?;
+    let (status_code, _body) = api_v1_db_schema(
+        Extension(ta2.agent.clone()),
+        axum::Json(vec![corro_tests::TEST_SCHEMA.into()]),
+    )
+    .await;
+    assert_eq!(status_code, StatusCode::OK);
+
+    for i in 1..=5_i64 {
+        let (status_code, _) = api_v1_transactions(
+            Extension(ta2.agent.clone()),
+            axum::Json(vec![Statement::WithParams(
+                "INSERT OR REPLACE INTO tests (id,text) VALUES (?,?)".into(),
+                vec![i.into(), "service-text".into()],
+            )]),
+        )
+        .await;
+        assert_eq!(status_code, StatusCode::OK);
+    }
+    let mut rows = get_rows(ta2.agent.clone(), vec![(Version(1)..=Version(5), None)]).await?;
+
+    let change6 = Change {
+        table: TableName("tests".into()),
+        pk: pack_columns(&vec![6i64.into()])?,
+        cid: ColumnName("text".into()),
+        val: "six".into(),
+        col_version: 1,
+        db_version: CrsqlDbVersion(6),
+        seq: CrsqlSeq(0),
+        site_id: actor_id.to_bytes(),
+        cl: 1,
+    };
+
+    let bad_change = Change {
+        table: TableName("tests".into()),
+        pk: pack_columns(&vec![6i64.into()])?,
+        cid: ColumnName("nonexistent".into()),
+        val: "six".into(),
+        col_version: 1,
+        db_version: CrsqlDbVersion(2),
+        seq: CrsqlSeq(1),
+        site_id: actor_id.to_bytes(),
+        cl: 1,
+    };
+
+    rows.push((
+        ChangeV1 {
+            actor_id,
+            changeset: Changeset::Full {
+                version: Version(1),
+                changes: vec![change6.clone(), bad_change],
+                seqs: CrsqlSeq(0)..=CrsqlSeq(1),
+                last_seq: CrsqlSeq(1),
+                ts: Default::default(),
+            },
+        },
+        ChangeSource::Sync,
+        Instant::now(),
+    ));
+
+    let res = process_multiple_changes(ta1.agent.clone(), ta1.bookie.clone(), rows).await;
+
+    assert!(res.is_ok());
+
+    // verify that correct versions were inserted
+    for i in 1..5 {
+        let conn = ta1.agent.pool().read().await?;
+        conn.prepare_cached("SELECT text from tests where id = ?")?
+            .query_row([i], |row| row.get::<_, String>(0))?;
+    }
+
+    let conn = ta1.agent.pool().read().await?;
+    let res = conn
+        .prepare_cached("SELECT text from tests where id = 6")?
+        .query_row([], |row| row.get::<_, String>(0));
+    assert!(res.is_err());
+    assert_eq!(res, Err(rusqlite::Error::QueryReturnedNoRows));
+
```

Comment:
Can you also check that the `db_version` inserted and the ones Corrosion associated with its internal version are correct? Possibly by trying to apply the bad change first.

I vaguely recall we do some db version arithmetics and we might need to subtract when we're rolling back a savepoint.